### PR TITLE
Add prefix to search and include search for name and job fields

### LINKED ISSUE #89: Add Name and Job prefix to Search Command
- add prefix to search command
- allow searching partial match of name and job

### PR DESCRIPTION
**Pull request by kaiyichen**
resolves #89 
- added name and job prefix to search
- able to search an applicant based on partial matching keywords of name or job

### CODE REVIEW

File: src/main/java/seedu/linkedout/logic/parser/SearchCommandParser.java [hunk 1]
```diff
@@ -15,18 +20,57 @@
...
+     *
      * @throws ParseException if the user input does not conform the expected format
      */
     public SearchCommand parse(String args) throws ParseException {
-        String trimmedArgs = args.trim();
-        if (trimmedArgs.isEmpty()) {
+
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_JOB);
+
+        if (!anyPrefixesPresent(argMultimap, PREFIX_NAME, PREFIX_JOB)
```
Comment (RyanCheungJF) [inline src/main/java/seedu/linkedout/logic/parser/SearchCommandParser.java:30]: This expression can be a bit over-complicated, if possible can split the 3 conditions into 3 different booleans with corresponding names. 

Written in pseudocode:

```suggestion
        boolean nameForCond1 = !anyPrefixes ...
        boolean nameForCond2 = !argMultimap...
        boolean nameForCond3 = !args.isEmpty()...
        if (cond1 || cond2 || cond3)
```

File: src/main/java/seedu/linkedout/logic/parser/SearchCommandParser.java [hunk 2]
```diff
@@ -15,18 +20,57 @@
...
+            return new SearchCommand(new NameContainsKeywordsPredicate(Arrays.asList(nameKeywords)));
+        } else {
+            String[] jobKeywords = getArrayOfKeywords(PREFIX_JOB, argMultimap);
+            return new SearchCommand(new JobContainsKeywordsPredicate(Arrays.asList(jobKeywords)));
+        }
+    }
 
-        return new SearchCommand(new NameContainsKeywordsPredicate(Arrays.asList(nameKeywords)));
+
+    private static boolean anyPrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
```
Comment (RyanCheungJF) [inline src/main/java/seedu/linkedout/logic/parser/SearchCommandParser.java:46]: Would be good to add a javadoc comment describing the method!

**Review by RyanCheungJF — COMMENTED** [review]
Tested on my own local machine and works well. Overall a great addition! There are some small stuff which could be changed to improve code quality.

File: src/main/java/seedu/linkedout/logic/commands/SearchCommand.java [hunk 3]
```diff
@@ -14,14 +17,18 @@
...
 
+    public static final String MESSAGE_CONSTRAINTS = "Keyword(s) should not be empty";
+
     public static final String MESSAGE_USAGE = COMMAND_WORD + ": Search an overview of a specific applicant"
-            + "specified by applicant's name (case-insensitive) and displays the applicant's information.\n"
-            + "Parameters: KEYWORD [MORE_KEYWORDS]...\n"
-            + "Example: " + COMMAND_WORD + "Steve";
+            + "Parameters: "
+            + PREFIX_NAME + "NAME "
+            + PREFIX_JOB + "Job "
```
[the remark this answers is not in the document]
Comment (zzhikai) [reply src/main/java/seedu/linkedout/logic/commands/SearchCommand.java:25]: I was thinking whether we want to have the message be [n/Name] [j/Job] but still mention that at least one of it is compulsory for the search to work, as judging from this message, it means both prefixes are compulsory.

**Review by zzhikai — COMMENTED** [review]
Great work on the search Command! Overall clean and well written code including the tests.

To be discussed during group meeting:

- Feasibility of combining the different prefix as some sort of search with filterS instead of just matching one prefix

Will review again and approve once @RyanCheungJF requested changes are done!

**Review by zzhikai — APPROVED** [review]
LGTM!

File: src/main/java/seedu/linkedout/logic/parser/SearchCommandParser.java [hunk 4]
```diff
@@ -15,18 +20,62 @@
...
+     *
      * @throws ParseException if the user input does not conform the expected format
      */
     public SearchCommand parse(String args) throws ParseException {
-        String trimmedArgs = args.trim();
-        if (trimmedArgs.isEmpty()) {
+
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_JOB);
+
+        boolean noPrefixesPresent = !anyPrefixesPresent(argMultimap, PREFIX_NAME, PREFIX_JOB);
```
Comment (RyanCheungJF) [inline src/main/java/seedu/linkedout/logic/parser/SearchCommandParser.java:30]: Apologies for requesting another change but would be good to rename in accordance to coding standards:
(Same for the other booleans used)

image


```suggestion
        boolean hasNoPrefixesPresent = !anyPrefixesPresent(argMultimap, PREFIX_NAME, PREFIX_JOB);
```

**Review by RyanCheungJF — CHANGES_REQUESTED** [review]
One minor last change, to keep the the naming of the booleans in line with the coding standards.

**Review by RyanCheungJF — APPROVED** [review]

**Pull request merged**